File: cmd/bt/history.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/BrokkAi/brokk-town/internal/town"
)

func historyCommand(ctx context.Context, dir string, fl *cliFlags) error {
	if *fl.repo == "" {
		return errors.New("history requires --repo OWNER/REPO")
	}
	if *fl.limit < 0 || *fl.limit > 100 {
		return errors.New("history --limit must be between 1 and 100, or omitted for 50")
	}
	if *fl.task != "" && (*fl.historyAfter != "" || *fl.limit != 0) {
		return errors.New("history --task cannot include pagination flags")
	}
	conn, err := requireService(ctx, dir)
	if err != nil {
		return err
	}
	body := map[string]any{"town": *fl.repo}
	if *fl.task != "" {
		body["task"] = *fl.task
		var result town.Task
		if err := request(ctx, conn, "POST", "/api/history", body, &result); err != nil {
			return err
		}
		if *fl.json {
			data, _ := json.MarshalIndent(result, "", "  ")
			fmt.Println(string(data))
			return nil
		}
		fmt.Printf("%s · %s · %s\n%s\n%s\n", result.ID, result.Stage, result.Updated.Format("2006-01-02"), result.Title, result.Detail)
		if result.MayoralDecision != "" {
			fmt.Println("Mayor decision:", result.MayoralDecision)
		}
		if result.Audit != nil {
			fmt.Println(result.Audit.Summary)
		}
		return nil
	}
	body["after"], body["limit"] = *fl.historyAfter, *fl.limit
	var page town.HistoryPage
	if err := request(ctx, conn, "POST", "/api/history", body, &page); err != nil {
		return err
	}
	if *fl.json {
		data, _ := json.MarshalIndent(page, "", "  ")
		fmt.Println(string(data))
		return nil
	}
	fmt.Printf("%s · %d archived tasks · active history retained for %d days\n", page.Town, page.Total, page.RetentionDays)
	for _, item := range page.Items {
		fmt.Printf("%s  %s  %s  %s\n", item.ID, item.Stage, item.Updated.Format("2006-01-02"), item.Title)
	}
	if page.Next != "" {
		fmt.Printf("Next page: --after %q\n", page.Next)
	}
	return nil
}

File: internal/web/history.js
const $ = (id) => document.querySelector(`#${id}`);

// History is read only on explicit navigation. Normal rendering uses its count.
export function historyPanel({ api, getTown }) {
  let version = 0, abort = null, busy = false, town = "", next = "";
  function controls() {
    $("history-newest").disabled = busy;
    $("history-older").disabled = busy || !next;
  }
  async function load(after = "", task = "") {
    if (busy || !town) return;
    busy = true; controls();
    const generation = version;
    abort = new AbortController();
    const signal = AbortSignal.any([abort.signal, AbortSignal.timeout(30000)]);
    $("history-error").textContent = "";
    $("history-status").textContent = task ? "Loading archived task…" : "Loading saved history…";
    try {
      const body = task ? { town, task } : { town, after, limit: 50 };
      const result = await api("/api/history", body, signal);
      if (generation !== version) return;
      if (task) {
        $("history-detail").textContent = [`${result.id} · ${result.stage}`, result.title, result.detail, result.mayoral_decision ? `Mayor decision: ${result.mayoral_decision}` : "", result.audit?.summary || ""].filter(Boolean).join("\n\n");
      } else {
        next = result.next || "";
        $("history-detail").textContent = "";
        $("history-summary").textContent = `${result.total} archived tasks. Completed tasks leave active history after ${result.retention_days} days; unfinished work stays active.`;
        const rows = result.items.map((item) => {
          const row = document.createElement("tr");
          for (const text of [item.id, item.title, item.stage, new Date(item.updated).toLocaleDateString()]) {
            const cell = document.createElement("td"); cell.textContent = text; row.append(cell);
          }
          const cell = document.createElement("td"), button = document.createElement("button");
          button.type = "button"; button.textContent = "Inspect"; button.onclick = () => load("", item.id);
          cell.append(button); row.append(cell); return row;
        });
        $("history-items").replaceChildren(...rows);
      }
      $("history-status").textContent = "Saved history. Reopened work returns to the active town automatically.";
    } catch (error) {
      if (generation === version) { $("history-error").textContent = signal.aborted ? "History lookup stopped. You can refresh to try again." : error.message; $("history-status").textContent = ""; }
    } finally { if (generation === version) { busy = false; controls(); } }
  }
  $("history-newest").onclick = () => load();
  $("history-older").onclick = () => load(next);
  $("history-close").onclick = () => $("history-dialog").close();
  $("history-dialog").addEventListener("close", () => { version++; abort?.abort(); busy = false; });
  $("town-history").onclick = () => {
    const selected = getTown(); if (!selected) return;
    version++; abort?.abort(); busy = false; next = ""; town = selected.id;
    $("history-town").textContent = town; $("history-items").replaceChildren(); $("history-detail").textContent = ""; $("history-summary").textContent = "";
    controls(); $("history-dialog").showModal(); void load();
  };
}
